tower-buffer/src/lib.rs

```diff
@@ -40,6 +43,54 @@ where
     worker: worker::Handle<T::Error>,
 }
 
+/// Buffer requests with a bounded buffer
+pub struct BufferLayer<E, S, Request> {
+    bound: usize,
+    executor: E,
+    _pd: PhantomData<(S, Request)>,
+}
+
+impl<S, Request> BufferLayer<DefaultExecutor, S, Request> {
+    pub fn new(bound: usize) -> Self {
+        BufferLayer {
+            bound,
+            executor: DefaultExecutor::current(),
+            _pd: PhantomData,
+        }
+    }
+}
+
+impl<E, S, Request> BufferLayer<E, S, Request>
+where
+    S: Service<Request>,
+    E: WorkerExecutor<S, Request>,
+{
+    pub fn with_executor(bound: usize, executor: E) -> Self {
+        BufferLayer {
+            bound,
+            executor,
+            _pd: PhantomData,
+        }
+    }
+}
+
+impl<E, S, Request> Layer<S, Request> for BufferLayer<E, S, Request>
+where
+    S: Service<Request>,
+    E: WorkerExecutor<S, Request>,
+{
+    type Response = S::Response;
+    type Error = Error<S::Error>;
+    type Service = Buffer<S, Request>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        match Buffer::with_executor(service, self.bound, &self.executor) {
+            Ok(b) => b,
+            Err(_) => panic!("Unable to spawn task on the provided executor"),
```

Comment:
This... is not good. Layering will happen at runtime.

This means that either `layer` has to return result, or it has to return a DOA service. The first is probably best.

Reply:
+1 on layer returning a result

Reply:
The alternative is to update `Buffer::new` to be infallible but if the spawn fails, then the first call to `poll_ready` returns Err, so a DOA Buffer instance.

Reply:
The problem with a DOA buffer instance is if you *then* wrap that with `Reconnect`, it would result in an infinite loop of `Reconnect` attempting to reconnect...